// ant-plus/build/blood-pressure-sensors.js
/// <reference path="../typings/tsd.d.ts"/>
/// <reference path="../typings/ant.d.ts"/>
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var Ant = require('./ant');
var Constants = Ant.Constants;
var Messages = Ant.Messages;
var BloodPressureSensor = (function (_super) {
    __extends(BloodPressureSensor, _super);
    function BloodPressureSensor(stick) {
        _super.call(this, stick);
        this.pageState = PageState.INIT_PAGE; // sets the state of the receiver - INIT, STD_PAGE, EXT_PAGE
        this.decodeDataCbk = this.decodeData.bind(this);
    }
    BloodPressureSensor.prototype.attach = function (channel, deviceID) {
        _super.prototype.attach.call(this, channel, 'receive', deviceID, BloodPressureSensor.deviceType, 0, 255, 8192);
        this.state = new HeartRateSensorState(deviceID);
    };
    BloodPressureSensor.prototype.decodeData = function (data) {
        if (data.readUInt8(Messages.BUFFER_INDEX_CHANNEL_NUM) !== this.channel) {
            return;
        }
        switch (data.readUInt8(Messages.BUFFER_INDEX_MSG_TYPE)) {
            case 78 /* MESSAGE_CHANNEL_BROADCAST_DATA */:
            case 79 /* MESSAGE_CHANNEL_ACKNOWLEDGED_DATA */:
            case 80 /* MESSAGE_CHANNEL_BURST_DATA */:
                {
                    if (this.deviceID === 0) {
                        this.write(Messages.requestMessage(this.channel, 81 /* MESSAGE_CHANNEL_ID */));
                    }
                    var page = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA);
                    if (this.pageState === PageState.INIT_PAGE) {
                        this.pageState = PageState.STD_PAGE; // change the state to STD_PAGE and allow the checking of old and new pages
                    }
                    else if ((page !== this.oldPage) || (this.pageState === PageState.EXT_PAGE)) {
                        this.pageState = PageState.EXT_PAGE; // set the state to use the extended page format
                        switch (page & ~HeartRateSensor.TOGGLE_MASK) {
                            case 1:
                                {
                                    //decode the cumulative operating time
                                    this.state.OperatingTime = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 1);
                                    this.state.OperatingTime |= data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 2) << 8;
                                    this.state.OperatingTime |= data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 3) << 16;
                                    this.state.OperatingTime *= 2;
                                    break;
                                }
                            case 2:
                                {
                                    //decode the Manufacturer ID
                                    this.state.ManId = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 1);
                                    //decode the 4 byte serial number
                                    this.state.SerialNumber = this.deviceID;
                                    this.state.SerialNumber |= data.readUInt16LE(Messages.BUFFER_INDEX_MSG_DATA + 2) << 16;
                                    this.state.SerialNumber >>>= 0;
                                    break;
                                }
                            case 3:
                                {
                                    //decode HW version, SW version, and model number
                                    this.state.HwVersion = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 1);
                                    this.state.SwVersion = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 2);
                                    this.state.ModelNum = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 3);
                                    break;
                                }
                            case 4:
                                {
                                    //decode the previous heart beat measurement time
                                    this.state.PreviousBeat = data.readUInt16LE(Messages.BUFFER_INDEX_MSG_DATA + 2);
                                    break;
                                }
                        }
                    }
                    // decode the last four bytes of the HRM format, the first byte of this message is the channel number
                    this.DecodeDefaultHRM(data.slice(Messages.BUFFER_INDEX_MSG_DATA + 4));
                    this.oldPage = page;
                }
                break;
            case 81 /* MESSAGE_CHANNEL_ID */:
                {
                    this.deviceID = data.readUInt16LE(Messages.BUFFER_INDEX_MSG_DATA);
                    this.transmissionType = data.readUInt8(Messages.BUFFER_INDEX_MSG_DATA + 3);
                    this.state.DeviceID = this.deviceID;
                }
                break;
        }
    };
    BloodPressureSensor.prototype.DecodeDefaultHRM = function (pucPayload) {
        // decode the measurement time data (two bytes)
        this.state.BeatTime = pucPayload.readUInt16LE(0);
        // decode the measurement count data
        this.state.BeatCount = pucPayload.readUInt8(2);
        // decode the measurement count data
        this.state.ComputedHeartRate = pucPayload.readUInt8(3);
        this.emit('hbdata', this.state);
    };
    BloodPressureSensor.deviceType = 18;
    BloodPressureSensor.TOGGLE_MASK = 0x80;
    return BloodPressureSensor;
})(Ant.AntPlusSensor);
exports.BloodPressureSensor = BloodPressureSensor;
